File: libjiffy/src/jiffy/storage/client/hash_table_listener.cpp
#include <thrift/transport/TTransportException.h>
#include "hash_table_listener.h"
#include "jiffy/storage/manager/detail/block_id_parser.h"
#include "../../utils/logger.h"

using namespace apache::thrift::transport;
using namespace jiffy::utils;

namespace jiffy {
namespace storage {

hash_table_listener::hash_table_listener(const std::string &path, const directory::data_status &status)
    : path_(path), status_(status), worker_(notifications_, controls_) {
  for (const auto &block: status_.data_blocks()) {
    auto t = block_id_parser::parse(block.block_names.back());
    block_ids_.push_back(t.id);
    listeners_.push_back(std::make_shared<block_listener>(t.host, t.notification_port, controls_));
    worker_.add_protocol(listeners_.back()->protocol());
  }
  worker_.start();
}

hash_table_listener::~hash_table_listener() {
  try {
    for (auto &listener: listeners_) {
      listener->disconnect();
    }
    worker_.stop();
  } catch (TTransportException &e) {
    LOG(log_level::info) << "Could not destruct: " << e.what();
  }
}

void hash_table_listener::subscribe(const std::vector<std::string> &ops) {
  for (size_t i = 0; i < listeners_.size(); i++) {
    listeners_[i]->subscribe(block_ids_[i], ops);
  }
}

void hash_table_listener::unsubscribe(const std::vector<std::string> &ops) {
  for (size_t i = 0; i < listeners_.size(); i++) {
    listeners_[i]->unsubscribe(block_ids_[i], ops);
  }
}

hash_table_listener::notification_t hash_table_listener::get_notification(int64_t timeout_ms) {
  return notifications_.pop(timeout_ms);
}

}
}

File: jiffy4j/core/src/main/java/jiffy/notification/HashTableListener.java
package jiffy.notification;

import java.io.Closeable;
import java.io.IOException;
import java.util.List;
import java.util.concurrent.TimeUnit;
import jiffy.directory.rpc_data_status;
import jiffy.directory.rpc_replica_chain;
import jiffy.storage.BlockNameParser;
import jiffy.storage.BlockNameParser.BlockMetadata;
import jiffy.notification.event.Control;
import jiffy.notification.event.Notification;
import org.apache.thrift.TException;
import org.apache.thrift.transport.TTransportException;

public class HashTableListener implements Closeable {

  private Mailbox<Notification> notifications;
  private Mailbox<Control> controls;
  private NotificationWorker worker;
  private Thread workerThread;
  private BlockListener[] listeners;
  private int[] blockIds;

  public HashTableListener(String path, rpc_data_status status) throws TTransportException {
    this.notifications = new Mailbox<>();
    this.controls = new Mailbox<>();
    this.blockIds = new int[status.data_blocks.size()];
    this.listeners = new BlockListener[status.data_blocks.size()];
    this.worker = new NotificationWorker(notifications, controls);
    int i = 0;
    for (rpc_replica_chain block: status.data_blocks) {
      BlockMetadata t = BlockNameParser.parse(block.block_ids.get(block.block_ids.size() - 1));
      blockIds[i] = t.getBlockId();
      listeners[i] = new BlockListener(t.getHost(), t.getServicePort(), controls);
      worker.addProtocol(listeners[i].getProtocol());
    }
    workerThread = new Thread(worker);
    workerThread.start();
  }

  @Override
  public void close() throws IOException {
    for (BlockListener listener: listeners) {
      listener.close();
    }
    worker.stop();
    try {
      workerThread.join();
    } catch (InterruptedException e) {
      throw new IOException(e);
    }
  }

  public void subscribe(List<String> ops) throws TException, InterruptedException {
    for (int i = 0; i < listeners.length; i++) {
      listeners[i].subscribe(blockIds[i], ops);
    }
  }

  public void unsubscribe(List<String> ops) throws TException, InterruptedException {
    for (int i = 0; i < listeners.length; i++) {
      listeners[i].unsubscribe(blockIds[i], ops);
    }
  }

  public Notification getNotification(int timeout, TimeUnit unit) throws InterruptedException {
    return notifications.poll(timeout, unit);
  }

  public Notification getNotification() throws InterruptedException {
    return notifications.take();
  }
}
